File: go/x402gateway/gateway.go
package x402gateway

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"net/http"
	"net/url"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// Options configure a Gateway. Zero values take the reference defaults.
type Options struct {
	SiteURL               string // required; canonical origin, no trailing slash
	SiteName              string // for the page; defaults to the hostname
	CoinPayAPIKey         string // a SCOPED CoinPay key (cp_live_…) with payments:create
	CoinPayBaseURL        string // default https://coinpayportal.com
	PayTo                 string // EVM address that receives the USDC
	PriceCents            float64
	Currency              string // default USD
	PassMinutes           int    // default 1440
	MaxDays               int    // default 30
	Header                string // default x-crawl-pass
	Path                  string // default /crawl
	OpenPaths             []string
	IsPaidAgent           func(userAgent string) bool
	DenyCIDRs             []string
	ChargeSpoofedBrowsers bool
	Exempt                func(*http.Request) bool
	Secret                string
	Training              []string
	Retrieval             []string
	Page                  func(PageContext) string
	Contact               string
	OnSale                func(Sale)
	HTTPClient            *http.Client
	Now                   func() int64 // unix seconds, for tests
}

// Sale is what OnSale receives.
type Sale struct {
	Payer      string
	Ref        string
	Token      string
	ExpiresAt  string
	UserAgent  string
	PriceCents float64
	Days       int
	TotalCents float64
	Currency   string
}

// Answer is a response the gateway wants sent instead of the site's.
type Answer struct {
	Status  int
	Headers http.Header
	Body    []byte
}

// Gateway sells crawl access to training crawlers, by the day, over x402.
type Gateway struct {
	o       Options
	Enabled bool
	secret  string
	open    []string
	denied  []CIDR
	buyURL  string
	coinpay CoinPay
}

// New builds a Gateway.
func New(o Options) (*Gateway, error) {
	o.SiteURL = strings.TrimRight(o.SiteURL, "/")
	if o.SiteURL == "" {
		return nil, errors.New("x402gateway: SiteURL is required")
	}
	if o.SiteName == "" {
		if u, err := url.Parse(o.SiteURL); err == nil && u.Hostname() != "" {
			o.SiteName = u.Hostname()
		} else {
			o.SiteName = o.SiteURL
		}
	}
	if o.CoinPayBaseURL == "" {
		o.CoinPayBaseURL = "https://coinpayportal.com"
	}
	o.CoinPayBaseURL = strings.TrimRight(o.CoinPayBaseURL, "/")
	if o.PriceCents == 0 || math.IsNaN(o.PriceCents) || math.IsInf(o.PriceCents, 0) {
		o.PriceCents = 100
	}
	if o.Currency == "" {
		o.Currency = "USD"
	}
	if o.PassMinutes <= 0 {
		o.PassMinutes = 1440
	}
	if o.MaxDays < 1 {
		o.MaxDays = 30
	}
	if o.Header == "" {
		o.Header = "x-crawl-pass"
	}
	o.Header = strings.ToLower(o.Header)
	if o.Path == "" {
		o.Path = "/crawl"
	}
	if o.Training == nil {
		o.Training = TrainingAgents
	}
	if o.Retrieval == nil {
		o.Retrieval = RetrievalAgents
	}
	if o.IsPaidAgent == nil {
		training := o.Training
		o.IsPaidAgent = func(ua string) bool { return IsTrainingAgent(ua, training) }
	}
	if o.Page == nil {
		o.Page = RenderPage
	}
	if o.Now == nil {
		o.Now = func() int64 { return time.Now().Unix() }
	}
	if o.HTTPClient == nil {
		o.HTTPClient = &http.Client{Timeout: 20 * time.Second}
	}
	g := &Gateway{o: o}
	g.Enabled = o.CoinPayAPIKey != "" && o.PayTo != ""
	g.secret = o.Secret
	if g.secret == "" {
		g.secret = o.CoinPayAPIKey
	}
	g.open = append([]string{"/robots.txt", o.Path, "/security.txt", "/.well-known/"}, o.OpenPaths...)
	g.denied = CompileCIDRs(o.DenyCIDRs)
	g.buyURL = o.SiteURL + o.Path
	g.coinpay = CoinPay{APIKey: o.CoinPayAPIKey, BaseURL: o.CoinPayBaseURL, Client: o.HTTPClient}
	return g, nil
}

// Options returns the normalised options.
func (g *Gateway) Options() Options { return g.o }

func (g *Gateway) money(cents float64) string {
	return strconv.FormatFloat(cents/100, 'f', 2, 64) + " " + g.o.Currency
}

func (g *Gateway) isOpen(path string) bool {
	for _, p := range g.open {
		if strings.HasSuffix(p, "/") {
			if strings.HasPrefix(path, p) {
				return true
			}
		} else if path == p {
			return true
		}
	}
	return false
}

var leadingInt = regexp.MustCompile(`^\s*([+-]?\d+)`)

func (g *Gateway) daysFrom(r *http.Request) int {
	m := leadingInt.FindStringSubmatch(r.URL.Query().Get("days"))
	if m == nil {
		return 1
	}
	n, err := strconv.Atoi(m[1])
	if err != nil || n < 1 {
		return 1
	}
	if n > g.o.MaxDays {
		return g.o.MaxDays
	}
	return n
}

// Offer is the offer for days terms: the same entries, days times the price.
func (g *Gateway) Offer(days int) Offer {
	if !g.Enabled {
		return Offer{X402Version: 2, Accepts: []Accept{}}
	}
	extra := ""
	if days > 1 {
		extra = fmt.Sprintf(" (%d × %d)", days, g.o.PassMinutes)
	}
	o, _ := BuildOffer(OfferOptions{
		PayTo: g.o.PayTo, PriceCents: g.o.PriceCents * float64(days), Resource: g.buyURL,
		Description: fmt.Sprintf("%d minutes of crawl access to %s%s", days*g.o.PassMinutes, g.o.SiteURL, extra),
	})
	return o
}

// PassInfo is the human half of a 402 body.
type PassInfo struct {
	Price   string `json:"price"`
	Minutes int    `json:"minutes"`
	Days    int    `json:"days"`
	Total   string `json:"total"`
	MaxDays int    `json:"maxDays"`
	Header  string `json:"header"`
	Buy     string `json:"buy"`
	BuyDays string `json:"buyDays"`
}

// Receipt is a 402 body: the offer, the pass terms, and why.
type Receipt struct {
	X402Version int      `json:"x402Version"`
	Accepts     []Accept `json:"accepts"`
	Pass        PassInfo `json:"pass"`
	Error       string   `json:"error,omitempty"`
}

func (g *Gateway) receipt(days int, errText string) Receipt {
	o := g.Offer(days)
	buy := g.buyURL
	if days > 1 {
		buy = fmt.Sprintf("%s?days=%d", g.buyURL, days)
	}
	return Receipt{
		X402Version: o.X402Version, Accepts: o.Accepts,
		Pass: PassInfo{
			Price: g.money(g.o.PriceCents), Minutes: g.o.PassMinutes, Days: days, Total: g.money(g.o.PriceCents * float64(days)),
			MaxDays: g.o.MaxDays, Header: g.o.Header, Buy: buy, BuyDays: g.buyURL + "?days=<n>",
		},
		Error: errText,
	}
}

func noStore(h http.Header) http.Header {
	h.Set("cache-control", "no-store")
	h.Set("vary", "Accept, User-Agent, X-Payment")
	return h
}

func (g *Gateway) jsonAnswer(body any, status int, extra map[string]string) *Answer {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	enc.SetIndent("", "  ")
	_ = enc.Encode(body)
	h := noStore(http.Header{})
	h.Set("content-type", "application/json; charset=utf-8")
	for k, v := range extra {
		h.Set(k, v)
	}
	return &Answer{Status: status, Headers: h, Body: bytes.TrimRight(buf.Bytes(), "\n")}
}

func (g *Gateway) htmlAnswer(body string, status int) *Answer {
	h := noStore(http.Header{})
	h.Set("content-type", "text/html; charset=utf-8")
	return &Answer{Status: status, Headers: h, Body: []byte(body)}
}

func (g *Gateway) pageCtx(days int) PageContext {
	return PageContext{
		Days: days, Total: g.money(g.o.PriceCents * float64(days)), SiteName: g.o.SiteName, SiteURL: g.o.SiteURL,
		BuyURL: g.buyURL, Price: g.money(g.o.PriceCents), Minutes: g.o.PassMinutes, MaxDays: g.o.MaxDays,
		Header: g.o.Header, Enabled: g.Enabled, Offer: g.Offer(1), Training: g.o.Training, Retrieval: g.o.Retrieval, Contact: g.o.Contact,
	}
}

// RobotsTxt is robots.txt with this gateway's lists and sales path. Zero fields of extra take the gateway's.
func (g *Gateway) RobotsTxt(extra RobotsOptions) string {
	if extra.SiteURL == "" {
		extra.SiteURL = g.o.SiteURL
	}
	if extra.Path == "" {
		extra.Path = g.o.Path
	}
	if extra.Training == nil {
		extra.Training = g.o.Training
	}
	if extra.Retrieval == nil {
		extra.Retrieval = g.o.Retrieval
	}
	return RobotsTxt(extra)
}

// Page is the sales page as HTML, for a site that mounts it on a route of its own.
func (g *Gateway) Page() string { return g.o.Page(g.pageCtx(1)) }

var bearer = regexp.MustCompile(`(?i)^Bearer\s+(cp_[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+)$`)

func (g *Gateway) passFrom(r *http.Request) string {
	if direct := header(r.Header, g.o.Header); direct != "" {
		return strings.TrimSpace(direct)
	}
	if m := bearer.FindStringSubmatch(header(r.Header, "authorization")); m != nil {
		return m[1]
	}
	return ""
}

func wantsHTML(accept string) bool { return strings.Contains(strings.ToLower(accept), "text/html") }

func iso(ts int64) string { return time.Unix(ts, 0).UTC().Format("2006-01-02T15:04:05.000Z") }

// Sell answers one request with the sale: a pass as the body of a 200, or a 402 with the offer.
func (g *Gateway) Sell(r *http.Request) *Answer {
	ua := header(r.Header, "user-agent")
	proofHeader := header(r.Header, "x-payment")
	asked := g.daysFrom(r)

	if proofHeader != "" {
		if !g.Enabled {
			return g.jsonAnswer(g.receipt(asked, "Payments are not switched on here."), 402, nil)
		}
		payment := DecodePayment(proofHeader)
		if payment == nil {
			return g.jsonAnswer(g.receipt(asked, "X-PAYMENT is not base64 JSON."), 402, nil)
		}
		unit := ExpectedFor(payment, g.Offer(1))
		if unit == nil {
			return g.jsonAnswer(g.receipt(asked, "Proof does not match an offered network."), 402, nil)
		}
		days := DaysPaid(PaidValueOf(payment), unit.Amount, g.o.MaxDays)
		if days == 0 {
			return g.jsonAnswer(g.receipt(asked, fmt.Sprintf("Pay a whole number of days: %s per day in the token's smallest unit, up to %d days. Add ?days=<n> to %s for the offer.", unit.Amount, g.o.MaxDays, g.buyURL)), 402, nil)
		}
		expected := ExpectedFor(payment, g.Offer(days))
		if expected == nil {
			expected = unit
		}
		term := int64(days) * int64(g.o.PassMinutes) * 60
		now := g.o.Now()
		result := g.coinpay.VerifyAndSettle(r.Context(), payment, *expected)

		var expiresAt int64
		replayed := false
		if result.OK {
			expiresAt = now + term
		} else if result.Replay {
			paid := g.coinpay.SettleAgain(r.Context(), payment)
			validBefore := ValidBeforeOf(payment)
			if paid && validBefore > 0 {
				expiresAt = now + term
				if validBefore+term < expiresAt {
					expiresAt = validBefore + term
				}
				replayed = true
			}
		}
		if expiresAt == 0 || expiresAt <= now {
			reason := result.Reason
			if reason == "" {
				reason = "Payment could not be settled."
			}
			return g.jsonAnswer(g.receipt(days, reason), 402, nil)
		}

		ref := NonceOf(payment)
		if ref == "" {
			ref = result.Ref
		}
		pass, err := MintPass(g.secret, ref, expiresAt, now)
		if err != nil {
			return g.jsonAnswer(g.receipt(days, "Payment could not be settled."), 402, nil)
		}
		expires := iso(pass.ExpiresAt)
		if g.o.OnSale != nil && !replayed {
			func() {
				defer func() { _ = recover() }() // accounting must never cost a buyer the pass
				g.o.OnSale(Sale{Payer: result.Payer, Ref: ref, Token: pass.Token, ExpiresAt: expires, UserAgent: ua,
					PriceCents: g.o.PriceCents, Days: days, TotalCents: g.o.PriceCents * float64(days), Currency: g.o.Currency})
			}()
		}
		body := struct {
			OK        bool   `json:"ok"`
			Pass      string `json:"pass"`
			ExpiresAt string `json:"expires_at"`
			Days      int    `json:"days"`
			Minutes   int    `json:"minutes"`
			Header    string `json:"header"`
			Replayed  bool   `json:"replayed"`
			Use       string `json:"use"`
		}{true, pass.Token, expires, days, days * g.o.PassMinutes, g.o.Header, replayed, fmt.Sprintf(`curl -H "%s: %s" %s/`, g.o.Header, pass.Token, g.o.SiteURL)}
		return g.jsonAnswer(body, 200, map[string]string{g.o.Header: pass.Token, g.o.Header + "-expires": expires})
	}

	if wantsHTML(header(r.Header, "accept")) {
		return g.htmlAnswer(g.o.Page(g.pageCtx(asked)), 402)
	}
	return g.jsonAnswer(g.receipt(asked, fmt.Sprintf("Payment required for training crawlers. Read %s for how.", g.buyURL)), 402, nil)
}

// Handle is the gate. Nil means "not for me, carry on".
func (g *Gateway) Handle(r *http.Request) *Answer {
	if len(g.denied) > 0 && InCIDRs(ClientIP(r), g.denied) {
		h := noStore(http.Header{})
		h.Set("content-type", "text/plain; charset=utf-8")
		return &Answer{Status: 403, Headers: h, Body: []byte("Not available from this network.\n")}
	}
	path := r.URL.Path
	if path == "" {
		path = "/"
	}
	if path == g.o.Path {
		return g.Sell(r)
	}
	if g.o.Exempt != nil && g.o.Exempt(r) {
		return nil
	}
	pays := g.o.IsPaidAgent(header(r.Header, "user-agent")) || (g.o.ChargeSpoofedBrowsers && IsSpoofedBrowser(r))
	if !pays {
		return nil
	}
	if g.isOpen(path) {
		return nil
	}
	if token := g.passFrom(r); token != "" && ReadPass(token, g.secret, g.o.Now()) != nil {
		return nil
	}
	return g.Sell(r)
}

// Write sends an Answer.
func (a *Answer) Write(w http.ResponseWriter) {
	for k, vs := range a.Headers {
		for _, v := range vs {
			w.Header().Add(k, v)
		}
	}
	w.Header().Set("content-length", strconv.Itoa(len(a.Body)))
	w.WriteHeader(a.Status)
	_, _ = w.Write(a.Body)
}

// Middleware wraps a handler: the gateway answers, or next does.
func (g *Gateway) Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if a := g.Handle(r); a != nil {
			a.Write(w)
			return
		}
		next.ServeHTTP(w, r)
	})
}

// RobotsHandler serves robots.txt.
func (g *Gateway) RobotsHandler(extra RobotsOptions) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("content-type", "text/plain; charset=utf-8")
		_, _ = w.Write([]byte(g.RobotsTxt(extra)))
	})
}

File: php/x402-gateway/src/Gateway.php
<?php
declare(strict_types=1);

namespace Profullstack\X402Gateway;

/**
 * A gateway that sells crawl access to training crawlers, by the day, over x402.
 *
 * handle(Request) returns a Response to send, or null to let the request through.
 * Options carry the reference's names in snake_case; see the README.
 */
final class Gateway
{
    private const NO_STORE = ['cache-control' => 'no-store', 'vary' => 'Accept, User-Agent, X-Payment'];

    public readonly string $siteUrl;
    public readonly string $siteName;
    public readonly string $header;
    public readonly string $path;
    public readonly bool $enabled;
    public readonly string $buyUrl;
    private string $coinpayApiKey;
    private string $coinpayBaseUrl;
    private string $payTo;
    private float $priceCents;
    private string $currency;
    private int $passMinutes;
    private int $maxDays;
    private array $open;
    private array $denied;
    private bool $chargeSpoofed;
    /** @var ?callable */
    private $exempt;
    private array $training;
    private array $retrieval;
    /** @var callable */
    private $isPaidAgent;
    private string $secret;
    /** @var callable */
    private $page;
    private string $contact;
    /** @var ?callable */
    private $onSale;
    /** @var callable */
    private $post;
    /** @var callable */
    private $now;

    public function __construct(array $o)
    {
        $siteUrl = rtrim((string) ($o['site_url'] ?? ''), '/');
        if ($siteUrl === '') {
            throw new \InvalidArgumentException('Gateway needs site_url');
        }
        $this->siteUrl = $siteUrl;
        $host = parse_url($siteUrl, PHP_URL_HOST);
        $this->siteName = (string) ($o['site_name'] ?? (is_string($host) && $host !== '' ? $host : $siteUrl));
        $this->coinpayApiKey = (string) ($o['coinpay_api_key'] ?? '');
        $this->coinpayBaseUrl = rtrim((string) ($o['coinpay_base_url'] ?? 'https://coinpayportal.com'), '/');
        $this->payTo = (string) ($o['pay_to'] ?? '');
        $price = $o['price_cents'] ?? 100;
        $this->priceCents = is_numeric($price) && is_finite((float) $price) ? (float) $price : 100.0;
        $this->currency = (string) ($o['currency'] ?? 'USD');
        $pm = $o['pass_minutes'] ?? 1440;
        $this->passMinutes = is_int($pm) && $pm > 0 ? $pm : 1440;
        $md = $o['max_days'] ?? 30;
        $this->maxDays = is_int($md) && $md >= 1 ? $md : 30;
        $h = strtolower((string) ($o['header'] ?? 'x-crawl-pass'));
        $this->header = $h === '' ? 'x-crawl-pass' : $h;
        $p = (string) ($o['path'] ?? '/crawl');
        $this->path = $p === '' ? '/crawl' : $p;
        $this->open = array_merge(['/robots.txt', $this->path, '/security.txt', '/.well-known/'], array_values($o['open_paths'] ?? []));
        $this->denied = Edge::compileCidrs($o['deny_cidrs'] ?? []);
        $this->chargeSpoofed = (bool) ($o['charge_spoofed_browsers'] ?? false);
        $this->exempt = $o['exempt'] ?? null;
        $this->training = array_values($o['training'] ?? Agents::TRAINING);
        $this->retrieval = array_values($o['retrieval'] ?? Agents::RETRIEVAL);
        $training = $this->training;
        $this->isPaidAgent = $o['is_paid_agent'] ?? fn (string $ua): bool => Agents::isTraining($ua, $training);
        $this->secret = (string) ($o['secret'] ?? '') !== '' ? (string) $o['secret'] : $this->coinpayApiKey;
        $this->page = $o['page'] ?? [Page::class, 'render'];
        $this->contact = (string) ($o['contact'] ?? '');
        $this->onSale = $o['on_sale'] ?? null;
        $this->post = $o['post'] ?? [X402::class, 'post'];
        $this->now = $o['now'] ?? fn (): int => time();
        $this->enabled = $this->coinpayApiKey !== '' && $this->payTo !== '';
        $this->buyUrl = $this->siteUrl . $this->path;
    }

    public function money(float $cents): string
    {
        return number_format($cents / 100, 2, '.', '') . ' ' . $this->currency;
    }

    public function price(): string
    {
        return $this->money($this->priceCents);
    }

    private function isOpen(string $path): bool
    {
        foreach ($this->open as $p) {
            if (str_ends_with($p, '/') ? str_starts_with($path, $p) : $path === $p) {
                return true;
            }
        }
        return false;
    }

    private function daysFrom(Request $r): int
    {
        if (!preg_match('/^\s*([+-]?\d+)/', (string) $r->query('days'), $m)) {
            return 1;
        }
        $n = strlen($m[1]) > 18 ? ($m[1][0] === '-' ? 0 : PHP_INT_MAX) : (int) $m[1];
        return $n < 1 ? 1 : min($n, $this->maxDays);
    }

    /** The offer for $days terms: the same entries, $days times the price. */
    public function offer(int $days = 1): array
    {
        if (!$this->enabled) {
            return ['x402Version' => 2, 'accepts' => []];
        }
        $extra = $days > 1 ? " ($days × {$this->passMinutes})" : '';
        return X402::buildOffer($this->payTo, $this->priceCents * $days, $this->buyUrl, ($days * $this->passMinutes) . " minutes of crawl access to {$this->siteUrl}$extra");
    }

    public function receipt(int $days = 1, array $extra = []): array
    {
        $body = $this->offer($days);
        $body['pass'] = [
            'price' => $this->price(), 'minutes' => $this->passMinutes, 'days' => $days, 'total' => $this->money($this->priceCents * $days),
            'maxDays' => $this->maxDays, 'header' => $this->header,
            'buy' => $days > 1 ? "{$this->buyUrl}?days=$days" : $this->buyUrl, 'buyDays' => "{$this->buyUrl}?days=<n>",
        ];
        return array_merge($body, $extra);
    }

    public function pageCtx(int $days = 1): array
    {
        return [
            'days' => $days, 'total' => $this->money($this->priceCents * $days), 'site_name' => $this->siteName, 'site_url' => $this->siteUrl,
            'buy_url' => $this->buyUrl, 'price' => $this->price(), 'minutes' => $this->passMinutes, 'max_days' => $this->maxDays,
            'header' => $this->header, 'enabled' => $this->enabled, 'offer' => $this->offer(), 'training' => $this->training,
            'retrieval' => $this->retrieval, 'contact' => $this->contact,
        ];
    }

    /** robots.txt with this gateway's lists and sales path. Keys of $extra: disallow, allow, sitemap, refused, comments. */
    public function robotsTxt(array $extra = []): string
    {
        return Robots::txt(
            $extra['site_url'] ?? $this->siteUrl, $extra['disallow'] ?? [], $extra['allow'] ?? [],
            array_key_exists('sitemap', $extra) ? $extra['sitemap'] : null, $extra['path'] ?? $this->path, $extra['refused'] ?? [],
            $extra['training'] ?? $this->training, $extra['retrieval'] ?? $this->retrieval, $extra['comments'] ?? []
        );
    }

    /** The sales page as HTML, for a site that mounts it on a route of its own. */
    public function page(): string
    {
        return ($this->page)($this->pageCtx());
    }

    private function json(array $body, int $status, array $extra = []): Response
    {
        return new Response($status, array_merge(['content-type' => 'application/json; charset=utf-8'], self::NO_STORE, $extra),
            json_encode($body, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE));
    }

    private function html(string $body, int $status): Response
    {
        return new Response($status, array_merge(['content-type' => 'text/html; charset=utf-8'], self::NO_STORE), $body);
    }

    private function passFrom(Request $r): ?string
    {
        $direct = $r->header($this->header);
        if ($direct !== null && $direct !== '') {
            return trim($direct);
        }
        return preg_match('/^Bearer\s+(cp_[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+)$/i', (string) $r->header('authorization'), $m) ? $m[1] : null;
    }

    /** Whether handling this request may call CoinPay (only a proof does). */
    public function needsIo(Request $r): bool
    {
        return $r->header('x-payment') !== null;
    }

    /** Answer one request with the sale: a pass as the body of a 200, or a 402 with the offer. */
    public function sell(Request $r): Response
    {
        $ua = (string) $r->header('user-agent');
        $proofHeader = $r->header('x-payment');
        $asked = $this->daysFrom($r);

        if ($proofHeader !== null && $proofHeader !== '') {
            if (!$this->enabled) {
                return $this->json($this->receipt($asked, ['error' => 'Payments are not switched on here.']), 402);
            }
            $payment = X402::decodePayment($proofHeader);
            if ($payment === null) {
                return $this->json($this->receipt($asked, ['error' => 'X-PAYMENT is not base64 JSON.']), 402);
            }
            $unit = X402::expectedFor($payment, $this->offer(1));
            if ($unit === null) {
                return $this->json($this->receipt($asked, ['error' => 'Proof does not match an offered network.']), 402);
            }
            $days = X402::daysPaid(X402::paidValueOf($payment), $unit['amount'], $this->maxDays);
            if ($days === 0) {
                return $this->json($this->receipt($asked, ['error' => "Pay a whole number of days: {$unit['amount']} per day in the token's smallest unit, up to {$this->maxDays} days. Add ?days=<n> to {$this->buyUrl} for the offer."]), 402);
            }
            $expected = X402::expectedFor($payment, $this->offer($days)) ?? $unit;
            $term = $days * $this->passMinutes * 60;
            $now = (int) ($this->now)();
            $result = X402::verifyAndSettle($payment, $expected, $this->coinpayApiKey, $this->coinpayBaseUrl, $this->post);

            $expiresAt = null;
            $replayed = false;
            if ($result['ok']) {
                $expiresAt = $now + $term;
            } elseif ($result['replay']) {
                $paid = X402::settleAgain($payment, $this->coinpayApiKey, $this->coinpayBaseUrl, $this->post);
                $validBefore = X402::validBeforeOf($payment);
                if ($paid && $validBefore !== null) {
                    $expiresAt = min($now + $term, $validBefore + $term);
                    $replayed = true;
                }
            }
            if ($expiresAt === null || $expiresAt <= $now) {
                return $this->json($this->receipt($days, ['error' => $result['reason'] ?? 'Payment could not be settled.']), 402);
            }

            $ref = X402::nonceOf($payment) ?? ($result['ref'] ?? null);
            $pass = Pass::mint($this->secret, $ref === null ? null : (string) $ref, $expiresAt, $now);
            $expires = gmdate('Y-m-d\TH:i:s', $pass['expires_at']) . '.000Z';
            if ($this->onSale !== null && !$replayed) {
                try {
                    ($this->onSale)([
                        'payer' => $result['payer'] ?? null, 'ref' => $ref, 'token' => $pass['token'], 'expires_at' => $expires, 'user_agent' => $ua,
                        'price_cents' => $this->priceCents, 'days' => $days, 'total_cents' => $this->priceCents * $days, 'currency' => $this->currency,
                    ]);
                } catch (\Throwable) {
                    // accounting must never cost a buyer the pass
                }
            }
            return $this->json([
                'ok' => true, 'pass' => $pass['token'], 'expires_at' => $expires, 'days' => $days, 'minutes' => $days * $this->passMinutes,
                'header' => $this->header, 'replayed' => $replayed, 'use' => "curl -H \"{$this->header}: {$pass['token']}\" {$this->siteUrl}/",
            ], 200, [$this->header => $pass['token'], "{$this->header}-expires" => $expires]);
        }

        if (str_contains(strtolower((string) $r->header('accept')), 'text/html')) {
            return $this->html(($this->page)($this->pageCtx($asked)), 402);
        }
        return $this->json($this->receipt($asked, ['error' => "Payment required for training crawlers. Read {$this->buyUrl} for how."]), 402);
    }

    /** The gate. Null means "not for me, carry on". */
    public function handle(Request $r): ?Response
    {
        if ($this->denied !== [] && Edge::inCidrs(Edge::clientIp($r), $this->denied)) {
            return new Response(403, array_merge(['content-type' => 'text/plain; charset=utf-8'], self::NO_STORE), "Not available from this network.\n");
        }
        $path = $r->path();
        if ($path === $this->path) {
            return $this->sell($r);
        }
        if ($this->exempt !== null && ($this->exempt)($r)) {
            return null;
        }
        $pays = ($this->isPaidAgent)((string) $r->header('user-agent')) || ($this->chargeSpoofed && Edge::isSpoofedBrowser($r));
        if (!$pays || $this->isOpen($path)) {
            return null;
        }
        $token = $this->passFrom($r);
        if ($token !== null && Pass::read($token, $this->secret, (int) ($this->now)()) !== null) {
            return null;
        }
        return $this->sell($r);
    }

    /**
     * Plain PHP: read the globals, and if the gateway answers, send it and return true.
     * Put `if ($gateway->handleGlobals()) exit;` at the top of your front controller.
     */
    public function handleGlobals(?array $server = null): bool
    {
        $answer = $this->handle(Request::fromGlobals($server));
        if ($answer === null) {
            return false;
        }
        $answer->send();
        return true;
    }
}
